src/lib/math/pcurves/pcurves_algos.h

```diff
@@ -0,0 +1,282 @@
+/*
+* (C) 2024,2025 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#ifndef BOTAN_PCURVES_ALGOS_H_
+#define BOTAN_PCURVES_ALGOS_H_
+
+#include <botan/types.h>
+#include <botan/internal/ct_utils.h>
+#include <span>
+#include <vector>
+
+namespace Botan {
+
+/**
+* Field inversion concept
+*
+* This concept checks if the FieldElement supports fe_invert2
+*/
+template <typename C>
+concept curve_supports_fe_invert2 = requires(const typename C::FieldElement& fe) {
+   { C::fe_invert2(fe) } -> std::same_as<typename C::FieldElement>;
+};
+
+/**
+* Field inversion
+*
+* Uses the specialized fe_invert2 if available, or otherwise the standard
+* (FLT-based) field inversion.
+*/
+template <typename C>
+inline constexpr auto invert_field_element(const typename C::FieldElement& fe) {
+   if constexpr(curve_supports_fe_invert2<C>) {
+      return C::fe_invert2(fe) * fe;
+   } else {
+      return fe.invert();
+   }
+}
+
+/**
+* Convert a projective point into affine
+*/
+template <typename C>
+inline constexpr auto to_affine(const typename C::ProjectivePoint& pt) {
+   // Not strictly required right? - default should work as long
+   // as (0,0) is identity and invert returns 0 on 0
+
+   if constexpr(curve_supports_fe_invert2<C>) {
+      const auto z2_inv = C::fe_invert2(pt.z());
+      const auto z3_inv = z2_inv.square() * pt.z();
+      return typename C::AffinePoint(pt.x() * z2_inv, pt.y() * z3_inv);
+   } else {
+      const auto z_inv = invert_field_element<C>(pt.z());
+      const auto z2_inv = z_inv.square();
+      const auto z3_inv = z_inv * z2_inv;
+      return typename C::AffinePoint(pt.x() * z2_inv, pt.y() * z3_inv);
+   }
```

Comment:
Do we even need the if-branch here? If I'm not missing something, the `invert_field_element` helper above should abstract this away entirely, no? Similarly in `to_affine_x`.

Reply:
Here computing the square of the inverse directly saves 1 field multiplication. And in `to_affine_x` saving 1 field mul and 1 field squaring.

Reply:
I'll add a comment

Reply:
But doesn't `invert_field_element<C>` produce inconsistent results then? Once the square of the inverse and once just the inverse, depending on the availability of the curve-specific implementation?

Reply:
Oh, right. Seems I should check my glasses. Now it makes sense. Thanks.